### ISSUE TITLE
Prefer SubFields vs. SelectionSet

### ISSUE THREAD

**Issue by joemcbride** [landmarkhw/Dapper.GraphQL#4]
Looks like a pretty cool library!

Not sure if it is well known, but the `ResolveFieldContext` does have a [`SubFields`](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Types/ResolveFieldContext.cs#L52) property which is the fields that should be requested.  It looks like that `Dapper.GraphQL` may have a bug in that even though a field is in the `SelectionSet`, it may not necessarily be allowed to be returned (due to directives mainly).

**Comment by dougrday** [comment]
Well, I think you'd know better than most :)  I'll update the source to use `SubFields` instead.

**Comment by benmccallum** [comment]
Nice work guys. Going to start using this today on a personal project and see how it goes. Thanks for logging the issue @joemcbride!

**Comment by benmccallum** [comment]
Another note for this change @dougrday , looks like when you upgrade to a later version of GraphQL that GraphQL.Type.Schema.ResolveType has been deprecated in favour of using the DependencyResolver property which is no longer a func but a concrete type implementing GraphQL.IDependencyResolver, so the docs/example will need to update similarly.

**Comment by joemcbride** [comment]
The main library is server agnostic so it does not have any dependencies on ASP.NET Core.  There is a `FuncDependencyResolver` provided for ease of transition.

```csharp
services.AddSingleton<IDependencyResolver>(
    s => new FuncDependencyResolver(type => s.GetRequiredService(type)));
```

**Comment by benmccallum** [comment]
Hey @joemcbride,

I've just come into a scenario as follows which reminded me of this ticket and your mention of using `context.SubFields` over `SelectionSet`. Right now I am using `SubFields`, but I'm needing to do paging and so I've setup the following:

A basic connection type, based off [GraphQL.Types.Relay.ConnectionType<TTo>](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Types/Relay/ConnectionType.cs) that just gives me a `totalCount` property for my UI. 
```csharp
public class BasicConnectionType<TTo> : ObjectGraphType<object>
        where TTo : IGraphType
{
    public BasicConnectionType()
    {
        Name = $"{typeof(TTo).GraphQLName()}Connection";
        Description = $"A basic (old-school paging) connection from an object to a list of objects of type `{typeof(TTo).GraphQLName()}`.";

        Field<IntGraphType>(
            name: "totalCount", 
            description: "A count of the total number of objects in this connection, ignoring pagination."
        );

        Field<ListGraphType<TTo>>(
            name: "items",
            description: $"A list of objects of type `{typeof(TTo).GraphQLName()}`."
        );
    }
}
```

In my root Query class:
```csharp
FieldAsync<BasicConnectionType<BookingType>>(...,
    resolve: async context =>
    {
        var fieldsRequested = context.SubFields.Keys; // ["totalCount", "items"]
        // I need whats under items, i.e. the fields on BookingType that were requested
    }
)
```

Basically I want to get those fields required on `BookingType` and pass them down to my DB query. That worked well when I was operating on the `ListGraphType<BookingType>>`, but now I'm wrapped in a `BasicConnectionType<TTo>` that `SubFields` list is `["totalCount", "items"]`. I can see what I want is under `context.SubFields.SelectionSet` but is it ok to use that given your comments above?

**Comment by benmccallum** [comment]
Hey @joemcbride, I'm currently re-visiting this TODO and wondering... is it possible now to get the SubFields of the SubFields yet? Doesn't look like it as `Field` doesn't have a SubFields prop yet. Should I create an issue in the GraphQL repo, or is this just flat out impossible?

If I just use SelectionSet, it seems like it still only gives you what was asked in the query, so are you saying that directives are the only graphql feature that should would make SubFields and SelectionSet different? Or are there other cases?
